pipeline/update_db.py

```diff
@@ -23,25 +35,63 @@ def get_db_connection():
     return eng
 
 
-def update_table(data, table, eng):
+def update_games(data, eng):
+    with eng.begin() as conn:
+        for _, row in data.iterrows():
+            stmt = f"SELECT * FROM {GAMES_TABLE} WHERE id = {row['id']};"
+            game = conn.execute(stmt).first()
+            # if game not in DB, insert all values and move on
+            if not game:
+                values = tuple(row[GAME_COLS])
+                stmt = f"INSERT INTO {GAMES_TABLE} VALUES {values};"
+                conn.execute(stmt)
+                continue
+            # if postponed/suspended, flag as date changed and nullify date
+            if row["date_changed"]:
+                stmt = (
+                    f"UPDATE {GAMES_TABLE} "
+                    f"SET date_changed = True, date = null "
+                    f"WHERE id = {row['id']};"
+                )
+                conn.execute(stmt)
+            # for games already in DB, update date/week/doubleheader
+            dt, wk, dh = row["date"], row["week"], row["dh"]
+            stmt = (
+                f"UPDATE {GAMES_TABLE} "
+                f"SET date = '{dt}', week = {wk}, dh = {dh} "
+                f"WHERE id = {row['id']};"
+            )
+            conn.execute(stmt)
+
+
+def update_outcomes(data, eng):
     with eng.begin() as conn:
-        # create temporary table with new data
-        data.to_sql("tmp", conn, index=False, if_exists="replace")
-        # remove those rows from games
-        stmt = f"DELETE FROM {table} WHERE id IN (SELECT id FROM tmp);"
-        conn.execute(stmt)
-        # insert new data into games
-        data.to_sql(table, conn, index=False, if_exists="append")
-        conn.execute("DROP TABLE tmp;")
+        for _, row in data.iterrows():
+            stmt = f"SELECT * FROM {OUTCOMES_TABLE} WHERE id = '{row['id']}';"
+            oc = conn.execute(stmt).first()
+            # if outcome not in DB, insert all values and move on
+            if not oc:
```

Comment:
I think I disagree here, if an outcome is the target of a wager, IMO it's best to have all possible outcomes in the table. This would also allow for changing betting odds in the future: the line would be associated with an outcome (as opposed to being two separate columns in `games`. Even if no one takes one side of a bet, the line would still need to be stored.

Reply:
That's a fair point. But if we're doing that, then we would want to put all possible future outcomes into the table as soon as games are scheduled. Inserting outcomes _after_ a game has already been played doesn't make much sense to me.

The use cases you mention seem like they're better handled at a different point in the process, and doing inserts at this step probably has more potential to cause bugs without much added value.

Reply:
If you want to insert all possible outcomes, that should be moved to an intermediate step after the `games` table is updated but before reporting the _actual_ outcome of completed games.

Reply:
> we would want to put all possible future outcomes into the table as soon as games are scheduled.

That's what's happening: outcomes are being generated for all games in the schedule data, including the ones in the future. So as soon as a game is inserted, its two outcomes are inserted as well. Those rows will be updated the day after the game completes. (Currently they're also updated with unnecessarily before they're played, but I'm gonna fix that now using the ORM.)

Does that make sense or am I missing something? Not sure I understand the value of separating the steps of inserting future outcomes and updating actual outcomes, since they're all in the same scraped outcomes data.

Reply:
:man_facepalming: yeah I just totally missed that part. OK this is making more sense now, lol!

I was thinking that in the future if we want to add other steps that are triggered based on final game results, it might be helpful to go ahead and split that out to a separate process. But for now this is doing exactly what we need (with the exception of unnecessary writes that will get optimized away once we add ORM logic).

Disregard my confusion!